plum/overload.py

```diff
@@ -0,0 +1,21 @@
+import sys
+from typing import Callable, TypeVar
+
+if sys.version_info >= (3, 11):  # pragma: specific no cover 3.7 3.8 3.9 3.10
+    from typing import get_overloads, overload
+else:  # pragma: specific no cover 3.11
+    from typing_extensions import get_overloads, overload
+
+from .function import Function
+
+__all__ = ["overload", "dispatch"]
+
+T = TypeVar("T", bound=Callable)
+
+
+def dispatch(f: T) -> T:
```

Comment:
```suggestion
def dispatch(f: Callable) -> Function:
```
Maybe mypy didn't catch this? I'm unsure

Reply:
Although this is technically correct, I believe this won't have the intended effect. Specifically, I tried changing the type stub to the following:

```python
from typing import Callable, TypeVar, overload

from plum import Function

__all__ = ["overload", "dispatch"]

T = TypeVar("T", bound=Callable)

def dispatch(f: Callable) -> Function: ...
```

Then `mypy` is happy with the following:


```python
from plum.overload import dispatch, overload


@overload
def f(x: int) -> int:
    return x


@overload
def f(x: str) -> str:
    return x


@dispatch
def f(x):
    pass


f(1)
f("1")
f(1.0)  # Should complain about this!!
```

Note that this is wrong, because it should complain about `f(1.0)`.

I believe that's what happening is the following. If you say that `dispatch` takes in `Callable` and gives back a `Function`, then `f(1.0)` will be matched against `Function.__call__`. `Function.__call__` takes in any arguments, so `f(1.0)` check out.

However, if you say that `dispatch` takes in `T` and gives back `T`, then `mypy` knows that it must correspond to the implementation in the file. The `@overload`s in the file correspond to the implementation of `f` in the file, so `mypy` knows that `f(1.0)` is wrong and then correctly complains.

Reply:
Hold on, I screwed up the test on my part! `mypy` actually does correctly complain about `f(1.0)` with the corrected type stub. I'm now a bit confused, because I don't exactly understand what's going on.

Reply:
I've now added tests for both `mypy` and `pyright`. It seems that `pyright` does indeed require the `T -> T` construction. If you'd like to experiment with it, you can run the linter assertion tests with

```bash
python check_linter_assertions.py tests/typechecked
```

from the repo root.